C union declarations are not yet supported, trigger ParseMatchError

A C program with a union declaration, such as the following, causes a `MatchError` and thus a crash:

```
union onion {
    float goat;
    char guitar;
};
```

It seems C and C++ unions are unsupported in general – probably for good reason! I imagine they're difficult to prove things about. However, in C++ mode, a "syntactically valid, but not supported" message is printed, with a blame pointed right at the declaration, which is a bit nicer.

<details>
<pre>
vct.parsers.err.ParseMatchError: A MatchError occurred while parsing. This likely indicates a missing case in a parser: [907 867 856 847 1636 1626 1620] (of class vct.antlr4.generated.CParser$StructOrUnionSpecifier0Context)
	at vct.parsers.AntlrParser.parseAntlrStreamOrThrow(AntlrParser.scala:155)
	at vct.parsers.AntlrParser.parseAntlrStream(AntlrParser.scala:196)
	at vct.parsers.AntlrParser.parseReader(AntlrParser.scala:205)
	at vct.parsers.Parser.$anonfun$parse$1(Parser.scala:34)
	at hre.io.Readable.read(Readable.scala:21)
	at hre.io.Readable.read$(Readable.scala:18)
	at hre.io.RWFile.read(RWFile.scala:7)
	at vct.parsers.Parser.parse(Parser.scala:34)
	at vct.parsers.parser.ColCParser.parseReader(ColCParser.scala:95)
	at vct.parsers.Parser.$anonfun$parse$1(Parser.scala:34)
	at hre.io.Readable.read(Readable.scala:21)
	at hre.io.Readable.read$(Readable.scala:18)
	at vct.options.types.PathOrStd$Path.read(PathOrStd.scala:56)
	at vct.parsers.Parser.parse(Parser.scala:34)
	at vct.main.stages.Parsing.$anonfun$run$2(Parsing.scala:139)
	at hre.progress.Progress$.$anonfun$map$2(Progress.scala:35)
	at hre.progress.task.AbstractTask.frame(AbstractTask.scala:166)
	at hre.progress.Progress$.$anonfun$map$1(Progress.scala:35)
	at scala.collection.Iterator$$anon$9.next(Iterator.scala:584)
	at scala.collection.immutable.List.prependedAll(List.scala:153)
	at scala.collection.immutable.List$.from(List.scala:684)
	at scala.collection.immutable.List$.from(List.scala:681)
	at scala.collection.SeqFactory$Delegate.from(Factory.scala:306)
	at scala.collection.immutable.Seq$.from(Seq.scala:42)
	at scala.collection.immutable.Seq$.from(Seq.scala:39)
	at scala.collection.IterableFactory$ToFactory.fromSpecific(Factory.scala:274)
	at scala.collection.IterableOnceOps.to(IterableOnce.scala:1310)
	at scala.collection.IterableOnceOps.to$(IterableOnce.scala:1310)
	at scala.collection.AbstractIterator.to(Iterator.scala:1300)
	at vct.main.stages.Parsing.run(Parsing.scala:140)
	at vct.main.stages.Parsing.run(Parsing.scala:79)
	at hre.stages.Stages.$anonfun$run$3(Stages.scala:104)
	at hre.stages.Stages.$anonfun$run$3$adapted(Stages.scala:101)
	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:576)
	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:574)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:933)
	at scala.collection.IterableOps$WithFilter.foreach(Iterable.scala:903)
	at hre.stages.Stages.$anonfun$run$1(Stages.scala:101)
	at hre.progress.task.NameSequenceTask.scope(NameSequenceTask.scala:16)
	at hre.progress.Progress$.stages(Progress.scala:47)
	at hre.stages.Stages.run(Stages.scala:98)
	at hre.stages.Stages.run$(Stages.scala:95)
	at hre.stages.StagesPair.run(Stages.scala:145)
	at vct.main.modes.Verify$.verifyWithOptions(Verify.scala:64)
	at vct.main.modes.Verify$.$anonfun$runOptions$3(Verify.scala:99)
	at scala.runtime.java8.JFunction0$mcI$sp.apply(JFunction0$mcI$sp.scala:17)
	at hre.util.Time$.logTime(Time.scala:23)
	at vct.main.modes.Verify$.runOptions(Verify.scala:99)
	at vct.main.Main$.runMode(Main.scala:107)
	at vct.main.Main$.$anonfun$runOptions$3(Main.scala:100)
	at scala.runtime.java8.JFunction0$mcI$sp.apply(JFunction0$mcI$sp.scala:17)
	at hre.middleware.Middleware$.using(Middleware.scala:78)
	at vct.main.Main$.$anonfun$runOptions$2(Main.scala:100)
	at scala.runtime.java8.JFunction0$mcI$sp.apply(JFunction0$mcI$sp.scala:17)
	at hre.io.Watch$.booleanWithWatch(Watch.scala:58)
	at vct.main.Main$.$anonfun$runOptions$1(Main.scala:100)
	at scala.runtime.java8.JFunction0$mcI$sp.apply(JFunction0$mcI$sp.scala:17)
	at hre.middleware.Middleware$.using(Middleware.scala:78)
	at vct.main.Main$.runOptions(Main.scala:95)
	at vct.main.Main$.main(Main.scala:50)
	at vct.main.Main.main(Main.scala)
[ERROR] !*!*!*!*!*!*!*!*!*!*!*!
[ERROR] ! VerCors has crashed !
[ERROR] !*!*!*!*!*!*!*!*!*!*!*!
</pre>
</details>

Version: 077f8b9842b25df833bd24e4a40cea43c34e57f8 (dev branch).

This issue was found by fuzzing (technically).

--- a/examples/technical/veymont/checkMainSyntaxAndWellFormedness/WaitStatement.pvl
+++ b/examples/technical/veymont/checkMainSyntaxAndWellFormedness/WaitStatement.pvl
@@ -1,5 +1,7 @@
-choreography Main() {
+class A {}
+
+choreography Main(A a) {
 	run {
-		wait null; // Not allowed
+		wait a; // Not allowed
 	}
 }

--- a/src/col/vct/col/ast/Node.scala
+++ b/src/col/vct/col/ast/Node.scala
@@ -448,40 +448,48 @@ final case class Instantiate[G](cls: Ref[G, Class[G]], out: Expr[G])(
 ) extends NormallyCompletingStatement[G]
     with ExpressionContainerStatement[G]
     with InstantiateImpl[G]
+sealed trait NonNullOperation[G] extends Node[G] with NonNullOperationImpl[G]
 final case class Wait[G](obj: Expr[G])(val blame: Blame[UnlockFailure])(
     implicit val o: Origin
 ) extends NormallyCompletingStatement[G]
     with ExpressionContainerStatement[G]
+    with NonNullOperation[G]
     with WaitImpl[G]
 final case class Notify[G](obj: Expr[G])(val blame: Blame[NotifyFailed])(
     implicit val o: Origin
 ) extends NormallyCompletingStatement[G]
     with ExpressionContainerStatement[G]
+    with NonNullOperation[G]
     with NotifyImpl[G]
 final case class Fork[G](obj: Expr[G])(val blame: Blame[ForkFailure])(
     implicit val o: Origin
 ) extends NormallyCompletingStatement[G]
     with ExpressionContainerStatement[G]
+    with NonNullOperation[G]
     with ForkImpl[G]
 final case class Join[G](obj: Expr[G])(val blame: Blame[JoinFailure])(
     implicit val o: Origin
 ) extends NormallyCompletingStatement[G]
     with ExpressionContainerStatement[G]
+    with NonNullOperation[G]
     with JoinImpl[G]
 final case class Lock[G](obj: Expr[G])(val blame: Blame[LockFailure])(
     implicit val o: Origin
 ) extends NormallyCompletingStatement[G]
     with ExpressionContainerStatement[G]
+    with NonNullOperation[G]
     with LockImpl[G]
 final case class Unlock[G](obj: Expr[G])(val blame: Blame[UnlockFailure])(
     implicit val o: Origin
 ) extends NormallyCompletingStatement[G]
     with ExpressionContainerStatement[G]
+    with NonNullOperation[G]
     with UnlockImpl[G]
 final case class Commit[G](obj: Expr[G])(val blame: Blame[CommitFailed])(
     implicit val o: Origin
 ) extends NormallyCompletingStatement[G]
     with ExpressionContainerStatement[G]
+    with NonNullOperation[G]
     with CommitImpl[G]
 final case class Fold[G](res: FoldTarget[G])(val blame: Blame[FoldFailed])(
     implicit val o: Origin
@@ -645,6 +653,7 @@ final case class Synchronized[G](obj: Expr[G], body: Statement[G])(
 )(implicit val o: Origin)
     extends CompositeStatement[G]
     with ControlContainerStatement[G]
+    with NonNullOperation[G]
     with SynchronizedImpl[G]
 @scopes[ParInvariantDecl]
 final case class ParInvariant[G](
@@ -2391,10 +2400,10 @@ final case class Then[G](value: Expr[G], post: Statement[G])(
 ) extends Expr[G] with ThenImpl[G]
 
 final case class Held[G](obj: Expr[G])(implicit val o: Origin)
-    extends Expr[G] with HeldImpl[G]
+    extends Expr[G] with NonNullOperation[G] with HeldImpl[G]
 final case class Committed[G](obj: Expr[G])(val blame: Blame[LockObjectNull])(
     implicit val o: Origin
-) extends Expr[G] with CommittedImpl[G]
+) extends Expr[G] with NonNullOperation[G] with CommittedImpl[G]
 final case class IdleToken[G](thread: Expr[G])(implicit val o: Origin)
     extends Expr[G] with IdleTokenImpl[G] with ResourceTerm[G]
 final case class JoinToken[G](thread: Expr[G])(implicit val o: Origin)

--- a/src/col/vct/col/ast/declaration/global/ClassImpl.scala
+++ b/src/col/vct/col/ast/declaration/global/ClassImpl.scala
@@ -6,6 +6,7 @@ import vct.col.ast.{
   Declaration,
   Expr,
   InstanceField,
+  TAnyClass,
   TByReferenceClass,
   TClass,
   TVar,
@@ -13,8 +14,10 @@ import vct.col.ast.{
   Variable,
 }
 import vct.col.ast.util.Declarator
+import vct.col.check.{CheckContext, CheckError, SupportNotAClass, TypeError}
 import vct.col.origin.Origin
 import vct.col.print._
+import vct.col.typerules.CoercionUtils
 import vct.col.util.AstBuildHelpers.tt
 
 trait ClassImpl[G] extends Declarator[G] {
@@ -91,4 +94,8 @@ trait ClassImpl[G] extends Declarator[G] {
       case Ctx.Java => layoutJava
       case _ => layoutPvl
     }
+
+  override def check(context: CheckContext[G]): Seq[CheckError] =
+    supports.map(s => (s, CoercionUtils.getCoercion(s, TAnyClass[G]())))
+      .collect { case (s, None) => SupportNotAClass(this, s) }
 }

--- a/src/col/vct/col/ast/expr/misc/OldImpl.scala
+++ b/src/col/vct/col/ast/expr/misc/OldImpl.scala
@@ -3,6 +3,7 @@ package vct.col.ast.expr.misc
 import vct.col.ast.{Old, Type}
 import vct.col.print.{Ctx, Doc, Precedence, Text}
 import vct.col.ast.ops.OldOps
+import vct.col.check.{CheckContext, CheckError, OldInPrecondition}
 
 trait OldImpl[G] extends OldOps[G] {
   this: Old[G] =>
@@ -18,4 +19,10 @@ trait OldImpl[G] extends OldOps[G] {
       case (_, Some(at)) =>
         Text("\\old[") <> ctx.name(at) <> "](" <> expr <> ")"
     }
+
+  override def check(context: CheckContext[G]): Seq[CheckError] =
+    if (context.inPreCondition)
+      Seq(OldInPrecondition(this))
+    else
+      Nil
 }

--- a/src/col/vct/col/ast/statement/terminal/NonNullOperationImpl.scala
+++ b/src/col/vct/col/ast/statement/terminal/NonNullOperationImpl.scala
@@ -1,0 +1,21 @@
+package vct.col.ast.statement.terminal
+
+import vct.col.ast.{Expr, NonNullOperation, TNull}
+import vct.col.ast.node.NodeImpl
+import vct.col.check.{CheckContext, CheckError, TypeErrorExplanation}
+
+trait NonNullOperationImpl[G] extends NodeImpl[G] {
+  this: NonNullOperation[G] =>
+
+  def obj: Expr[G]
+
+  override def check(context: CheckContext[G]): Seq[CheckError] =
+    obj.t match {
+      case TNull() =>
+        Seq(TypeErrorExplanation(
+          obj,
+          "This expression is statically known to be null but it may not be null",
+        ))
+      case _ => Nil
+    }
+}

--- a/src/col/vct/col/check/Check.scala
+++ b/src/col/vct/col/check/Check.scala
@@ -165,6 +165,11 @@ sealed trait CheckError {
             "The endpoint referenced in the outer expression here...",
           context(inner) -> "...differs from the endpoint referenced here",
         )
+      case SupportNotAClass(cls, support) =>
+        Seq(
+          context(cls) ->
+            s"This class cannot extend or implement the type $support since it is not a class"
+        )
     }): _*)
 
   def subcode: String
@@ -281,6 +286,9 @@ case class OnlyInChannelInvariant(expr: Node[_]) extends CheckError {
 case class InconsistentEndpointExprNesting(outer: Node[_], inner: Node[_])
     extends CheckError {
   val subcode = "inconsistentEndpointExprNesting"
+}
+case class SupportNotAClass(cls: Node[_], support: Type[_]) extends CheckError {
+  val subcode = "supportNotAClass"
 }
 
 case object CheckContext {

--- a/src/col/vct/col/check/Check.scala
+++ b/src/col/vct/col/check/Check.scala
@@ -170,6 +170,8 @@ sealed trait CheckError {
           context(cls) ->
             s"This class cannot extend or implement the type $support since it is not a class"
         )
+      case OldInPrecondition(expr) =>
+        Seq(context(expr) -> s"\\old may not be used in a precondition")
     }): _*)
 
   def subcode: String
@@ -289,6 +291,9 @@ case class InconsistentEndpointExprNesting(outer: Node[_], inner: Node[_])
 }
 case class SupportNotAClass(cls: Node[_], support: Type[_]) extends CheckError {
   val subcode = "supportNotAClass"
+}
+case class OldInPrecondition(expr: Node[_]) extends CheckError {
+  val subcode = "oldInPrecondition"
 }
 
 case object CheckContext {

--- a/src/parsers/antlr4/LangCPPLexer.g4
+++ b/src/parsers/antlr4/LangCPPLexer.g4
@@ -181,6 +181,8 @@ fragment Universalcharactername:
 	'\\u' Hexquad
 	| '\\U' Hexquad Hexquad;
 
+Identifier: Identifiernondigit (Identifiernondigit | DIGIT)*;
+
 fragment Identifiernondigit: NONDIGIT | Universalcharactername;
 
 fragment NONDIGIT: [a-zA-Z_];
@@ -301,9 +303,6 @@ Whitespace: [ \t]+ -> skip;
 
 Newline: ('\r' '\n'? | '\n') -> skip;
 
-
-mode DEFAULT_MODE;
-Identifier: Identifiernondigit (Identifiernondigit | DIGIT)*;
 
 ExtraAt: ('\n'|'\r\n') [ \t\u000C]* '@' {inBlockSpec}? -> skip;
 

--- a/src/parsers/antlr4/LangCPPLexer.g4
+++ b/src/parsers/antlr4/LangCPPLexer.g4
@@ -11,12 +11,14 @@ channels {
   LINE_DIRECTIVE_CHANNEL
 }
 
+
 // lexer tokens needed for the SpecParser
 VAL_INLINE: EOF EOF;
 VAL_ASSERT: 'assert';
 VAL_PACKAGE: 'package';
 VAL_BOOL: EOF EOF;
 
+mode DEFAULT_MODE;
 IntegerLiteral:
 	DecimalLiteral Integersuffix?
 	| OctalLiteral Integersuffix?

--- a/src/parsers/vct/parsers/transform/CToCol.scala
+++ b/src/parsers/vct/parsers/transform/CToCol.scala
@@ -242,8 +242,11 @@ case class CToCol[G](
             _,
           ) =>
         CStructDeclaration(name.map(convert), convert(declarations))
+      case StructOrUnionSpecifier0(union @ StructOrUnion1(_), _, _, _, _) =>
+        ??(union)
       case StructOrUnionSpecifier1(StructOrUnion0(_), name) =>
         CStructSpecifier(convert(name))
+      case StructOrUnionSpecifier1(union @ StructOrUnion1(_), _) => ??(union)
     }
 
   def convert(

